client/lib/signup/step-actions.js

```diff
@@ -171,16 +171,22 @@ export function createSiteWithCart(
 		'import' === flowName ? normalizeImportUrl( getNuxUrlInputValue( state ) ) : '';
 	const importEngine = 'import' === flowName ? getSelectedImportEngine( state ) : '';
 
+	// flowName isn't always passed in
+	const flowToCheck = flowName || lastKnownFlow;
+
 	if ( importingFromUrl ) {
 		newSiteParams.blog_name = importingFromUrl;
 		newSiteParams.find_available_url = true;
 		newSiteParams.options.nux_import_engine = importEngine;
+	} else if ( ! siteUrl && isDomainStepSkippable( flowToCheck ) ) {
+		newSiteParams.blog_name =
+			get( user.get(), 'username', siteTitle ) || siteType || getSiteVertical( state );
```

Comment:
As mentioned, you could try pulling it out of `state.signup.dependencyStore.username`

Reply:
If and when we move the signup step to the end of the flow however, then it might become redundant anyway, so the fallbacks you have are 👍